crate_universe/src/metadata/cargo_tree_resolver.rs

```diff
@@ -66,71 +107,195 @@ impl TreeResolver {
+                        "Error running `cargo tree --target={}` (host = '{}'), manifest path '{}'",
                         target_triple,
+                        host_triple,
                         manifest_path.display()
                     )
                 })?;
-            if !output.status.success() {
-                eprintln!("{}", String::from_utf8_lossy(&output.stdout));
-                eprintln!("{}", String::from_utf8_lossy(&output.stderr));
-                bail!(format!("Failed to run cargo tree: {}", output.status))
+                if !output.status.success() {
+                    eprintln!("{}", String::from_utf8_lossy(&output.stdout));
+                    eprintln!("{}", String::from_utf8_lossy(&output.stderr));
+                    bail!(format!("Failed to run cargo tree: {}", output.status))
+                }
+
+                // Replicate outputs for any de-duplicated platforms
+                for host_plat in cargo_host_triples[host_triple].iter() {
+                    for target_plat in cargo_target_triples[&target_triple].iter() {
+                        stdouts
+                            .entry((*host_plat).clone())
+                            .or_default()
+                            .insert((*target_plat).clone(), output.stdout.clone());
+                    }
+                }
             }
-            stdouts.insert(target_triple, output.stdout);
         }
 
         Ok(stdouts)
     }
 
+    // The use of this wrapper should __never__ escape this class.
+    #[cfg(target_family = "windows")]
+    fn create_rustc_wrapper_impl(output_dir: &Path) -> Result<PathBuf> {
+        let wrapper = output_dir.join("cargo_tree_rustc_wrapper.bat");
+        std::fs::write(
+            &wrapper,
+            include_str!(concat!(
+                env!("CARGO_MANIFEST_DIR",),
+                "/src/metadata/cargo_tree_rustc_wrapper.bat"
+            )),
+        )
+        .context("Failed to write rustc wrapper")?;
+
+        Ok(wrapper)
+    }
+
+    // The use of this wrapper should __never__ escape this class.
+    #[cfg(target_family = "unix")]
+    fn create_rustc_wrapper_impl(output_dir: &Path) -> Result<PathBuf> {
+        let wrapper = output_dir.join("cargo_tree_rustc_wrapper.sh");
+        std::fs::write(
+            &wrapper,
+            include_str!(concat!(
+                env!("CARGO_MANIFEST_DIR",),
+                "/src/metadata/cargo_tree_rustc_wrapper.sh"
+            )),
+        )
+        .context("Failed to write rustc wrapper")?;
+
+        use std::os::unix::fs::PermissionsExt;
+
+        let perms = std::fs::Permissions::from_mode(0o700);
+        std::fs::set_permissions(&wrapper, perms)
+            .context("Failed to modify permissions of rustc wrapper")?;
+
+        Ok(wrapper)
+    }
+
+    /// Create a wrapper for `rustc` which can intercept commands used to identify the host
```

Comment:
Can you link to https://github.com/rust-lang/cargo/issues/14527?

Reply:
Done!